# Manipulação e Agregação de Dados - Exercícios

Resolva os exercícios propostos abaixo. Os exercícios usam um dataset de reclamações/solicitações registradas por cidadãos. O detaset foi derivado de dados do [Portal de Dados Abertos da Prefeitura de Curitiba](https://www.curitiba.pr.gov.br/DADOSABERTOS/).

## Leitura e análise inicial dos dados

Inicie um DataFrame a partir do arquivo `2017-02-01_156_-_Base_de_Dados_sample.csv`. 

**Dica:** Caso o arquivo não seja separado por vírgulas, o Pandas não conseguirá reconhecer os campos adequadamente. Você precisará fornecer o parâmetro `sep=';'` para indicar o separador correto (neste caso, o `;`).

**Dica:** Aplicativos modernos tendem a armazenar arquivos no formato UTF-8, mas é comum encontrar arquivos codificados em outros formatos. Caso você tenha problemas para ler o arquivo, utilize o parâmetro `encoding='latin-1'` para selecionar o encoding correto (*latin-1* neste exemplo).

In [2]:
# Resposta:
import pandas as pd

df = pd.read_csv("../data/2017-02-01_156_-_Base_de_Dados_sample.csv", encoding='latin-1', sep=';')

df

SOLICITACAO         TIPO                               ORGAO  \
0         6669771  SOLICITAÇÃO  INSTITUTO DAS CIDADES INTELIGENTES   
1         6718351  SOLICITAÇÃO  INSTITUTO DAS CIDADES INTELIGENTES   
2         6702371  SOLICITAÇÃO  INSTITUTO DAS CIDADES INTELIGENTES   
3         6718692  SOLICITAÇÃO  INSTITUTO DAS CIDADES INTELIGENTES   
4         6704503  SOLICITAÇÃO  INSTITUTO DAS CIDADES INTELIGENTES   
...           ...          ...                                 ...   
7008      6690744  SOLICITAÇÃO  INSTITUTO DAS CIDADES INTELIGENTES   
7009      6716157  SOLICITAÇÃO  INSTITUTO DAS CIDADES INTELIGENTES   
7010      6689584  SOLICITAÇÃO  INSTITUTO DAS CIDADES INTELIGENTES   
7011      6661976  SOLICITAÇÃO  INSTITUTO DAS CIDADES INTELIGENTES   
7012      6734925  SOLICITAÇÃO  INSTITUTO DAS CIDADES INTELIGENTES   

            DATA   HORARIO                            ASSUNTO  \
0     09/11/2016  21:02:44   ABORDAGEM SOCIAL DE RUA - ADULTO   
1     27/12/2016  00:40:04   ABORDAGEM SOCIAL DE RUA - ADULTO   
2     07/12/2016  14:09:42  ABORDAGEM SOCIAL DE RUA - CRIANÇA   
3     27/12/2016  13:28:27   ABORDAGEM SOCIAL DE RUA - ADULTO   
4     09/12/2016  01:28:16   ABORDAGEM SOCIAL DE RUA - ADULTO   
...          ...       ...                                ...   
7008  28/11/2016  13:04:25                           DRENAGEM   
7009  22/12/2016  10:11:15                           TRÂNSITO   
7010  26/11/2016  21:03:07  ABORDAGEM SOCIAL DE RUA - CRIANÇA   
7011  03/11/2016  15:54:31             NOVO MOBILIÁRIO URBANO   
7012  11/01/2017  20:10:14   ABORDAGEM SOCIAL DE RUA - ADULTO   

                                       SUBDIVISAO  \
0            PESSOAS/FAMÍLIAS EM DESABRIGO NA RUA   
1                            PERDIDA/DESORIENTADA   
2                           ALCOOLIZADAS/DROGADAS   
3            PESSOAS/FAMÍLIAS EM DESABRIGO NA RUA   
4                          DORMINDO/CAÍDAS NA RUA   
...                                           ...   
7008  LIMPEZA E DESOBSTRUÇÃO DE CAIXA DE CAPTAÇÃO   
7009                VEÍCULO ESTACIONADO - ESQUINA   
7010           DESENVOLVENDO ATIVIDADES INFORMAIS   
7011                              PONTO DE ÔNIBUS   
7012                         PERDIDA/DESORIENTADA   

                                              DESCRICAO  \
0     ABORDAGEM SOCIAL DE RUA - ADULTO - PESSOAS/FAM...   
1     ABORDAGEM SOCIAL DE RUA - ADULTO - PERDIDA/DES...   
2     ABORDAGEM SOCIAL DE RUA - CRIANÇA - ALCOOLIZAD...   
3     ABORDAGEM SOCIAL DE RUA - ADULTO - PESSOAS/FAM...   
4     ABORDAGEM SOCIAL DE RUA - ADULTO - DORMINDO/CA...   
...                                                 ...   
7008  DRENAGEM - LIMPEZA E DESOBSTRUÇÃO DE CAIXA DE ...   
7009           TRÂNSITO - VEÍCULO ESTACIONADO - ESQUINA   
7010  ABORDAGEM SOCIAL DE RUA - CRIANÇA - DESENVOLVE...   
7011           NOVO MOBILIÁRIO URBANO - PONTO DE ÔNIBUS   
7012  ABORDAGEM SOCIAL DE RUA - ADULTO - PERDIDA/DES...   

                             LOGRADOURO_ASS         BAIRRO_ASS  \
0                        ANDRE DE BARROS, 0             CENTRO   
1                             NILO CAIRO, 0             CENTRO   
2                   VISCONDE DE NACAR, 1210             CENTRO   
3                       MARECHAL DEODORO, 0             CENTRO   
4                       MARECHAL DEODORO, 0             CENTRO   
...                                     ...                ...   
7008  DEPUTADO HEITOR ALENCAR FURTADO, 4000  CIDADE INDUSTRIAL   
7009                  SAO PAULO DA CRUZ, 50             JUVEVE   
7010                          DO ROSARIO, 0      SAO FRANCISCO   
7011                   COMENDADOR FRANCO, 0        PRADO VELHO   
7012        PRESIDENTE AFFONSO CAMARGO, 330    JARDIM BOTANICO   

                 REGIONAL_ASS MEIO_RESPOSTA  \
0     Unidade Regional Matriz        NENHUM   
1     Unidade Regional Matriz        NENHUM   
2     Unidade Regional Matriz        NENHUM   
3     Unidade Regional Matriz        NENHUM   
4     Unid

Visualize algumas linhas de dados:

In [3]:
# Resposta:
df.head(5)

SOLICITACAO         TIPO                               ORGAO        DATA  \
0      6669771  SOLICITAÇÃO  INSTITUTO DAS CIDADES INTELIGENTES  09/11/2016   
1      6718351  SOLICITAÇÃO  INSTITUTO DAS CIDADES INTELIGENTES  27/12/2016   
2      6702371  SOLICITAÇÃO  INSTITUTO DAS CIDADES INTELIGENTES  07/12/2016   
3      6718692  SOLICITAÇÃO  INSTITUTO DAS CIDADES INTELIGENTES  27/12/2016   
4      6704503  SOLICITAÇÃO  INSTITUTO DAS CIDADES INTELIGENTES  09/12/2016   

    HORARIO                            ASSUNTO  \
0  21:02:44   ABORDAGEM SOCIAL DE RUA - ADULTO   
1  00:40:04   ABORDAGEM SOCIAL DE RUA - ADULTO   
2  14:09:42  ABORDAGEM SOCIAL DE RUA - CRIANÇA   
3  13:28:27   ABORDAGEM SOCIAL DE RUA - ADULTO   
4  01:28:16   ABORDAGEM SOCIAL DE RUA - ADULTO   

                             SUBDIVISAO  \
0  PESSOAS/FAMÍLIAS EM DESABRIGO NA RUA   
1                  PERDIDA/DESORIENTADA   
2                 ALCOOLIZADAS/DROGADAS   
3  PESSOAS/FAMÍLIAS EM DESABRIGO NA RUA   
4                DORMINDO/CAÍDAS NA RUA   

                                           DESCRICAO           LOGRADOURO_ASS  \
0  ABORDAGEM SOCIAL DE RUA - ADULTO - PESSOAS/FAM...       ANDRE DE BARROS, 0   
1  ABORDAGEM SOCIAL DE RUA - ADULTO - PERDIDA/DES...            NILO CAIRO, 0   
2  ABORDAGEM SOCIAL DE RUA - CRIANÇA - ALCOOLIZAD...  VISCONDE DE NACAR, 1210   
3  ABORDAGEM SOCIAL DE RUA - ADULTO - PESSOAS/FAM...      MARECHAL DEODORO, 0   
4  ABORDAGEM SOCIAL DE RUA - ADULTO - DORMINDO/CA...      MARECHAL DEODORO, 0   

  BAIRRO_ASS             REGIONAL_ASS MEIO_RESPOSTA  \
0     CENTRO  Unidade Regional Matriz        NENHUM   
1     CENTRO  Unidade Regional Matriz        NENHUM   
2     CENTRO  Unidade Regional Matriz        NENHUM   
3     CENTRO  Unidade Regional Matriz        NENHUM   
4     CENTRO  Unidade Regional Matriz        NENHUM   

                                          OBSERVACAO SEXO BAIRRO_CIDADAO  \
0               NÃO SOUBE INFORMAR NUMERAÇÃO PREDIAL    M            NaN   
1  NÃO SOUBE INFORMAR NUMERAÇÃO PREDIAL, RELATA E...    M            NaN   
2                                                NaN    F            NaN   
3                                                NaN    M            NaN   
4              NÃO SOUBE INFORMAR NUMERAÇÃO PREDIAL.    M            NaN   

  REGIONAL_CIDADAO   DATA_NASC TIPO_CIDADAO               ORGAO_RESP  \
0              NaN         NaN      CIDADÃO  FUNDAÇÃO DE AÇÃO SOCIAL   
1              NaN         NaN      CIDADÃO  FUNDAÇÃO DE AÇÃO SOCIAL   
2              NaN         NaN      CIDADÃO  FUNDAÇÃO DE AÇÃO SOCIAL   
3              NaN         NaN      CIDADÃO  FUNDAÇÃO DE AÇÃO SOCIAL   
4              NaN  04/06/1997      CIDADÃO  FUNDAÇÃO DE AÇÃO SOCIAL   

                                      RESPOSTA_FINAL  
0  Abordagem realizada. Pessoa foi orientada quan...  
1  EM ABORDAGEM REALIZADA, NÃO FOI ENCONTRADA A P...  
2  - EM ABORDAGEM REALIZADA, NÃO FOI ENCONTRADA A...  
3  A pessoa da solicitação já foi abordada nesta ...  
4  A pessoa da solicitação já faz parte da rede d...

Use o método `df.describe(include="all")` para exibir informações sobre as colunas.

In [4]:
# Resposta:
df.describe(include="all")

SOLICITACAO         TIPO                               ORGAO  \
count   7.013000e+03         7013                                7013   
unique           NaN            7                                   5   
top              NaN  SOLICITAÇÃO  INSTITUTO DAS CIDADES INTELIGENTES   
freq             NaN         5812                                6649   
mean    6.702377e+06          NaN                                 NaN   
std     2.668504e+04          NaN                                 NaN   
min     6.659367e+06          NaN                                 NaN   
25%     6.680119e+06          NaN                                 NaN   
50%     6.700369e+06          NaN                                 NaN   
75%     6.722191e+06          NaN                                 NaN   
max     6.763349e+06          NaN                                 NaN   

              DATA   HORARIO ASSUNTO                SUBDIVISAO  \
count         7013      7013    7013                      7013   
unique          92      6438     148                       388   
top     05/12/2016  18:27:15  COLETA  MANUTENÇÃO DE LUMINÁRIAS   
freq           163         4    1633                       969   
mean           NaN       NaN     NaN                       NaN   
std            NaN       NaN     NaN                       NaN   
min            NaN       NaN     NaN                       NaN   
25%            NaN       NaN     NaN                       NaN   
50%            NaN       NaN     NaN                       NaN   
75%            NaN       NaN     NaN                       NaN   
max            NaN       NaN     NaN                       NaN   

                                            DESCRICAO  \
count                                            7013   
unique                                            482   
top     ILUMINAÇÃO PÚBLICA - MANUTENÇÃO DE LUMINÁRIAS   
freq                                              969   
mean                                              NaN   
std                                               NaN   
min                                               NaN   
25%                                               NaN   
50%                                               NaN   
75%                                               NaN   
max                                               NaN   

                         LOGRADOURO_ASS BAIRRO_ASS             REGIONAL_ASS  \
count                              7013       7013                     7011   
unique                             5612         75                       10   
top     PRESIDENTE AFFONSO CAMARGO, 330     CENTRO  Unidade Regional Matriz   
freq                                 39        436                     1572   
mean                                NaN        NaN                      NaN   
std                                 NaN        NaN                      NaN   
min                                 NaN        NaN                      NaN   
25%                                 NaN        NaN                      NaN   
50%                                 NaN        NaN                      NaN   
75%                                 NaN        NaN                      NaN   
max                                 NaN        NaN                      NaN   

       MEIO_RESPOSTA                             OBSERVACAO  SEXO  \
count           7013                                   3786  7013   
unique             4                                   2585     2   
top            EMAIL  NÃO SOUBE INFORMAR NUMERAÇÃO PREDIAL.     F   
freq            3750                                     88  3613   
mean             NaN                                    NaN   NaN   
std              NaN                                    NaN   NaN   
min              NaN                                    NaN   NaN   
25%              NaN                                    NaN   NaN   
50%              NaN                                    NaN   NaN   
75%            

Crie um novo DataFrame chamado `df_elogios` contendo apenas as linhas cujos valores na coluna `TIPO` são iguais a 'ELOGIO'. Este DataFrame deve ter as colunas TIPO, ASSUNTO, e BAIRRO_CIDADAO.

In [5]:
# Resposta:
df_elogios = df[df.TIPO=="ELOGIO"]

df_elogios

SOLICITACAO    TIPO                               ORGAO        DATA  \
124       6660527  ELOGIO  INSTITUTO DAS CIDADES INTELIGENTES  01/11/2016   
457       6692174  ELOGIO  INSTITUTO DAS CIDADES INTELIGENTES  29/11/2016   
573       6666083  ELOGIO       SECRETARIA MUNICIPAL DA SAÚDE  07/11/2016   
981       6683367  ELOGIO  INSTITUTO DAS CIDADES INTELIGENTES  22/11/2016   
1023      6727772  ELOGIO  INSTITUTO DAS CIDADES INTELIGENTES  06/01/2017   
...           ...     ...                                 ...         ...   
6700      6726248  ELOGIO  INSTITUTO DAS CIDADES INTELIGENTES  05/01/2017   
6759      6701368  ELOGIO  INSTITUTO DAS CIDADES INTELIGENTES  06/12/2016   
6897      6716219  ELOGIO  INSTITUTO DAS CIDADES INTELIGENTES  22/12/2016   
6944      6752901  ELOGIO  INSTITUTO DAS CIDADES INTELIGENTES  24/01/2017   
7001      6683174  ELOGIO  INSTITUTO DAS CIDADES INTELIGENTES  22/11/2016   

       HORARIO                             ASSUNTO  \
124   22:02:50                              COLETA   
457   11:57:07  MOTORISTAS, COBRADORES E PORTEIROS   
573   15:30:00                       OUVIDORIA SMS   
981   10:06:35                SINALIZAÇÃO VERTICAL   
1023  17:19:08                      POSTO DE SAÚDE   
...        ...                                 ...   
6700  17:52:29                      POSTO DE SAÚDE   
6759  21:27:41                  ESPORTE ESTUDANTIL   
6897  11:05:04  MOTORISTAS, COBRADORES E PORTEIROS   
6944  14:52:52       UNIDADE DE PRONTO ATENDIMENTO   
7001  08:56:08  MOTORISTAS, COBRADORES E PORTEIROS   

                                       SUBDIVISAO  \
124                       RECLAMAÇÃO (DOMICILIAR)   
457          ELOGIO PELO BOM DESEMPENHO DA FUNÇÃO   
573                              ATENDIMENTO 0800   
981                      PLACA DE ÁREA MONITORADA   
1023  RECURSOS HUMANOS - ATENDIMENTO PROFISSIONAL   
...                                           ...   
6700  RECURSOS HUMANOS - ATENDIMENTO PROFISSIONAL   
6759                      INFORMAÇÕES SOBRE JOGOS   
6897         ELOGIO PELO BOM DESEMPENHO DA FUNÇÃO   
6944  RECURSOS HUMANOS - ATENDIMENTO PROFISSIONAL   
7001         ELOGIO PELO BOM DESEMPENHO DA FUNÇÃO   

                                              DESCRICAO  \
124                    COLETA - RECLAMAÇÃO (DOMICILIAR)   
457   MOTORISTAS, COBRADORES E PORTEIROS - ELOGIO PE...   
573                    OUVIDORIA SMS - ATENDIMENTO 0800   
981     SINALIZAÇÃO VERTICAL - PLACA DE ÁREA MONITORADA   
1023  POSTO DE SAÚDE - RECURSOS HUMANOS - ATENDIMENT...   
...                                                 ...   
6700  POSTO DE SAÚDE - RECURSOS HUMANOS - ATENDIMENT...   
6759       ESPORTE ESTUDANTIL - INFORMAÇÕES SOBRE JOGOS   
6897  MOTORISTAS, COBRADORES E PORTEIROS - ELOGIO PE...   
6944  UNIDADE DE PRONTO ATENDIMENTO - RECURSOS HUMAN...   
7001  MOTORISTAS, COBRADORES E PORTEIROS - ELOGIO PE...   

                    LOGRADOURO_ASS         BAIRRO_ASS  \
124   VISCONDE DE GUARAPUAVA, 5045              BATEL   
457                      PARANÁ, 0      SANTA CANDIDA   
573            AUGUSTO SEVERO, 295     ALTO DA GLORIA   
981                   PARANÁ, 3654          BACACHERI   
1023                RIO PELOTAS, 0        BAIRRO ALTO   
...                            ...                ...   
6700   ARTHUR MARTINS FRANCO, 5516  CIDADE INDUSTRIAL   
6759  GUILHERME WALTER LOWRY, 1655            UBERABA   
6897                 TIRADENTES, 0             CENTRO   
6944                  PARANÁ, 3654          BACACHERI   
7001                BRASILIA, 6676         NOVO MUNDO   

                      REGIONAL_ASS MEIO_RESPOSTA  \
124        Unidade Regional Matriz         EMAIL   
457     Unidade Regional Boa Vista      TELEFONE   
573        Unidade Regional Matriz      TELEFONE   
981     Unidade Regional Boa Vista         EMAIL   
1023    Unidade Regional Boa Vista      TELEFONE   
...                            ...           ...   
6700          Unidade Regional CIC   

Quais são os 5 bairros com mais elogios? Agrupe pelo `BAIRRO_CIDADAO` e ordene o resultado para responder.

In [6]:
# Resposta:
df_grouped = df_elogios.groupby("BAIRRO_CIDADAO").size().sort_values(ascending=False).head(5)

df_grouped

BAIRRO_CIDADAO
SITIO CERCADO        12
CIDADE INDUSTRIAL    12
CAJURU               11
BOQUEIRAO             6
BAIRRO ALTO           5
dtype: int64

Crie dois novos DataFrames, um contendo os registros de pessoas do sexo masculino e outra do sexo feminino. Use estes DataFrames para responder:

Qual é a proporção de reclamações entre homens e mulheres? (total de reclamações de homens dividido pelo total de reclamações de mulheres)

In [ ]:
# Resposta:

masc_select = df["SEXO"] == "M"
fem_select = df["SEXO"] == "F"

df_masc = df[masc_select]
df_fem = df[fem_select]

reclamacoes_masc = len(df_masc[df_masc.TIPO=="RECLAMAÇÃO"])
reclamacoes_fem = len(df_fem[df_fem.TIPO=="RECLAMAÇÃO"])

proporcao_reclamacoes = reclamacoes_masc / reclamacoes_fem

print("Proporcao reclamacoes:",proporcao_reclamacoes)


Proporcao reclamacoes: 0.5081081081081081


Concatene os dois DataFrames criados acima para criar um DataFrame chamado `df_todos`. Verifique se este DataFrame tem o mesmo número de linhas do DataFrame lido do CSV.

In [ ]:
# Resposta:

df_todos = df_masc + df_fem

print("df_todos len:",len(df_todos), "df len:", len(df))

df_todos len: 7013 df len: 7013


Crie um novo DataFrame chamado `df_sexo` contendo a contagem de registros para cada sexo. A coluna com a contagem deve se chamar `contagem`.

In [ ]:
# Resposta:
df_sexo = df.groupby("SEXO").size().reset_index(name="contagem")

df_sexo

SEXO  contagem
0    F      3613
1    M      3400

Crie um novo DataFrame chamado `df_sexo_tipo` contendo a contagem de registros para cada sexo e tipo. A coluna com a contagem deve se chamar `total`.

In [ ]:
# Resposta:

df_sexo_tipo = df.groupby(["SEXO", "TIPO"]).size().reset_index(name="total")

df_sexo_tipo

SEXO         TIPO  total
0     F     DENÚNCIA     31
1     F       ELOGIO     65
2     F   INFORMAÇÃO     25
3     F   RECADASTRO     79
4     F   RECLAMAÇÃO    555
5     F  SOLICITAÇÃO   2851
6     F     SUGESTÃO      7
7     M     DENÚNCIA     13
8     M       ELOGIO     34
9     M   INFORMAÇÃO     11
10    M   RECADASTRO     91
11    M   RECLAMAÇÃO    282
12    M  SOLICITAÇÃO   2961
13    M     SUGESTÃO      8

Faça uma junção (join ou merge) dos DataFrames `df_sexo, df_sexo_tipo` para criar um novo DataFrame chamado `df_juncao` contendo as contagens por sexo e também por sexo/tipo.

In [ ]:
# Resposta:
df_juncao = pd.merge(df_sexo, df_sexo_tipo)

df_juncao

SEXO  contagem         TIPO  total
0     F      3613     DENÚNCIA     31
1     F      3613       ELOGIO     65
2     F      3613   INFORMAÇÃO     25
3     F      3613   RECADASTRO     79
4     F      3613   RECLAMAÇÃO    555
5     F      3613  SOLICITAÇÃO   2851
6     F      3613     SUGESTÃO      7
7     M      3400     DENÚNCIA     13
8     M      3400       ELOGIO     34
9     M      3400   INFORMAÇÃO     11
10    M      3400   RECADASTRO     91
11    M      3400   RECLAMAÇÃO    282
12    M      3400  SOLICITAÇÃO   2961
13    M      3400     SUGESTÃO      8

Crie uma nova coluna no DataFrame `df_juncao` chamada `proporcao` contendo a porcentagem de homens e mulheres que fizeram cada tipo de registro. Responda: qual sexo faz mais elogios? Qual faz mais solicitações? Qual faz mais reclamações?

In [ ]:
# Resposta:
df_juncao["proporcao"] = (df_juncao["total"] /df_juncao["contagem"]) * 100

df_juncao

SEXO  contagem         TIPO  total  proporcao
0     F      3613     DENÚNCIA     31   0.858013
1     F      3613       ELOGIO     65   1.799059
2     F      3613   INFORMAÇÃO     25   0.691946
3     F      3613   RECADASTRO     79   2.186549
4     F      3613   RECLAMAÇÃO    555  15.361196
5     F      3613  SOLICITAÇÃO   2851  78.909493
6     F      3613     SUGESTÃO      7   0.193745
7     M      3400     DENÚNCIA     13   0.382353
8     M      3400       ELOGIO     34   1.000000
9     M      3400   INFORMAÇÃO     11   0.323529
10    M      3400   RECADASTRO     91   2.676471
11    M      3400   RECLAMAÇÃO    282   8.294118
12    M      3400  SOLICITAÇÃO   2961  87.088235
13    M      3400     SUGESTÃO      8   0.235294

No conjunto de dados analisado, individuos do sexo feminino fizeram mais elogios, os 65 elogios feitos, são quase o dobro dos 34 feitos por indivíduos do sexo masculino. O mesmo padrão se repete para reclamações, onde as mulheres fizeram 555 reclamações e os homens 282. Para as solicitações, a proporção é mais equilibrada, com 2961 solicitações feitas por homens e 2851 feitas por mulheres.
